### ISSUE TITLE
Suggestion to allow more robust state providers

### ISSUE THREAD

**Issue by jtsmith0107** [ITV/scala-pact#41]
I've been looking into this repo for one of my own projects, and have found a need for more flexibility in how state providers are handled by the runner. I've looked at making this change for my own purposes. Perhaps there would be interest in making this change in master.

State providers currently do direct string comparison to parse to different handlers. Given example
```scala
providerStates := Seq(
  ("Resource with ID 1234 exists", (key: String) => {
    println("Injecting key 1234 into the database...")
    // Do some work to ensure the system under test is
    // in an appropriate state before verification
    true
  })
)
```
Limitation is specific resource 1234 must be agreed by both ends before hand. 

My suggestion is to change direct string comparison ( str1 == str2 ) to pattern matching. This can be done by simply making ProviderState a `String => Boolean` type instead of Seq[(String, String => Boolean)].

This allows some more language features for parsing strings. Example of my use:
```scala

// Regex matcher
val Resource = """Resource:([a-z]+)\{([^\{\}]+)\}""".r

providerStates := {
  case key: String if key.startsWith("Resource with ID 1234 exists") =>
    println("Injecting key 1234 into the database...")
    // Do some work to ensure the system under test is
    // in an appropriate state before verification

    true
  case Resource("user", id: String) => // Given Sample:  'Resource:User:id:1234'
   // Request to setup user with id
    true
 case "Resource with ID 4321 exists"
    true

}
```
Note we have ability for regex and boolean statements. Direct matching is still supported. Regex matching allows me to pass user id to the provider state logic to, for example, make http request to my server to create user with provided id before executing the test.

Main concern is this is a breaking change, so updating the library will incur errors for existing provider state definitions.

If interested I can put up a PR for this.

**Comment by davesmith00000** [comment]
Hello!

It's an interesting suggestion. When I was putting this together initially it did feel a bit rubbish and remember thinking "I'll come back to this later when I have more time..." - famous last words.

It is the way it is because it's a thoughtless straight interpretation of the Pact spec. There's a lookup key to match against the "given" field which is a string in the Pact spec ...which I then thought might be handy to have in your setup function ...and then maybe you should be able to indicate whether your setup succeeded or not hence:

`Set[(LookupKey, KeyAgainForReference => Successful?)]`

I'll try and give this some more thought, however two things immediately spring to mind:

1. This change doesn't need to be a breaking change, it can be additive, we just need to add an extra field that we try to use before or after the existing one. Maybe with a view to deprecating the old one at some point but no rush since people are using it.
2. `String => Boolean` aka `Function1[String, Boolean]` is fine but it's a total function working against unknowable input (could be any string right?), and it doesn't allow us to then indicate whether or not we managed to find a corresponding provider state match. We could force everyone to add an `case _ => false` clause but that's a bit sucky.

How about something like:

```scala
val providerStateMatcher = SettingKey[PartialFunction[String, Boolean]]("providerStateMatcher", "A pattern matching function enabled provider setup.")
```

...so that if the partial function fails to match we can catch the fact and report it?

Thoughts?

**Comment by jtsmith0107** [comment]
So to the point of maintaining a reference to the key, it shouldn't be a problem since we know what we want to match at the time of running the script. If it matches, we know the value. The Seq[String, String => Boolean] pattern seems redundant to me.

```scala
case key: String if key == "User with ID 1234" => 
  // use key reference
```
A default matcher could be added by just chaining PartialFunctions with provided orElse function after the user has defined providerState. So...

```scala
val defaultCase = { case _ => false} 
providerState orElse defaultCase
``` 
For avoiding breaking change, that's a good suggestion. That should be fully possible.

**Comment by davesmith00000** [comment]
Yes I think we are in agreement. Let me know if you plan on having a go at doing the work won't you?

**Comment by davesmith00000** [comment]
This was released in version 2.1.2.
